
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Edoardi Giacopuzzi">
      
      
      
        <link rel="prev" href="../input-reuse-step1/">
      
      
        <link rel="next" href="../tower-monitoring/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.17">
    
    
      
        <title>main parameters - nf-regenie-gwas</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.bcfcd587.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue" data-md-color-accent="blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#main-parameters" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="nf-regenie-gwas" class="md-header__button md-logo" aria-label="nf-regenie-gwas" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            nf-regenie-gwas
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              main parameters
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/HTGenomeAnalysisUnit/nf-pipeline-regenie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../quick-start/" class="md-tabs__link">
        
  
    
  
  Quick start

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../running-modes/" class="md-tabs__link">
          
  
  Running mode

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../input-phenotype-file/" class="md-tabs__link">
          
  
  Input files

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
  Configuration

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../outputs/" class="md-tabs__link">
        
  
    
  
  Output files

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../additional-notes/" class="md-tabs__link">
        
  
    
  
  Additional notes

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="nf-regenie-gwas" class="md-nav__button md-logo" aria-label="nf-regenie-gwas" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    nf-regenie-gwas
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/HTGenomeAnalysisUnit/nf-pipeline-regenie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../quick-start/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick start
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Running mode
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Running mode
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../running-modes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    single model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../running-modes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    multi projects
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../running-modes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    multi models
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Input files
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Input files
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-phenotype-file/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Phenotype / Covariate file
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-full-data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Full genetic dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-indep-snps/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Independent QCed SNPs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-ld-panel/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD panel (optional)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-rare-variants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inputs for rare variants tests
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-models-table/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Models table (multi models run)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-projects-table/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Project table (multi projects run)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-conditional-analysis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Conditional / interaction analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-reuse-step1/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Re-use step1 predictions
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Configuration
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    main parameters
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    main parameters
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Input files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multi-models-run" class="md-nav__link">
    <span class="md-ellipsis">
      Multi-models run
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multi-projects-run" class="md-nav__link">
    <span class="md-ellipsis">
      Multi-projects run
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conditional-and-interaction-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Conditional and interaction analysis
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tower-monitoring/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    tower monitoring
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../hpc-profile/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    define HPC profile
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../parallelization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    optimize parallelization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../top-hits-annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    top hits annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../full-params-list/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    full list of params
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../outputs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Output files
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../additional-notes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Additional notes
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Input files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multi-models-run" class="md-nav__link">
    <span class="md-ellipsis">
      Multi-models run
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multi-projects-run" class="md-nav__link">
    <span class="md-ellipsis">
      Multi-projects run
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conditional-and-interaction-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Conditional and interaction analysis
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="main-parameters">Main parameters</h1>
<p>A template for preparing a configuration file for a new project is provided in <code>templates/single_project.conf</code>. You can copy this file and edit it to adjust the parameters for your analysis.</p>
<p>Here is a list of the main parameters you need to adjust for a new analysis:</p>
<ul>
<li>
<p>Set <code>project</code> to a custom string defining your project ID. Note that max 50 chars are allowed and no special characters. This will be used in reports and a folder will be created with the same name to store all results.</p>
</li>
<li>
<p>Set <code>chromosomes</code> to represent the list of chromosome to be included in the analysis. You can use a comma-separated list of chromosome numbers, a range like <code>1-22</code> or a mix like <code>1,4,11-18</code>.</p>
</li>
<li>
<p>You can eventually restrict the analysis to a specific genomic range, or a specific list of SNP IDs or gene IDs using the <code>regenie_range</code>, <code>regenie_extract_snps</code> and <code>regenie_extract_genes</code> options. This requires <code>step2_gwas_split</code> and <code>step2_rarevar_split</code> to be set to <code>false</code>. Keep in mind that in this case you must ensure that the provided SNPs/genes/region are present in the genotype dataset you provided as input and in the specified chromosomes. Otherwise, the pipeline will fail due to SNPs remaining for step2 analysis.</p>
</li>
<li>
<p>Set <code>genotypes_build</code> to the build of your genotype data, either hg19 or hg38.</p>
</li>
<li>
<p>Set <code>step2_gwas_chunk_size</code> and <code>step2_rarevar_chunk_size</code> according to the size of your dataset. These control how the dataset is split for step2 analysis. Teh default values usually work fine, but you can increase/decrease them if you have very large or small datasets. Keep in mind that a small chunk size may result in a very large amount of parallel jobs. By default the pipeline job submission rate is limited to 200 concurrent jobs. The total number of jobs will be <em>N_SNPs/gwas_chunk_size</em> for GWAS and <em>N_genes/rare_chunk_size</em> for rare variant analysis.</p>
</li>
<li>
<p>For GWAS analysis, set <code>regenie_gwas_test</code> to the type of model you want to test, either 'additive', 'dominant' or 'recessive'.</p>
</li>
<li>
<p>For rare variant test, set <code>rarevars_vc_test</code> as a comma-separated list of rare variants tests to perform among those accepted by regenie: skat, skato, skato-acat, acatv, acato, acato-full. You can also set <code>rarevars_joint_test</code> to perform a joint test across burden masks / bins. Possible values are: minp, acat, sbat.</p>
</li>
<li>
<p>Set <code>regenie_gwas_min_mac</code> and <code>regenie_rarevar_min_mac</code> to control the min allowed MAC for variants tested in either GWAS or rare variant analysis. Variants with MAC below this threshold will be excluded from the analysis.</p>
</li>
<li>
<p>Set <code>annotation_min_log10p</code> to the min value ( -log10(pval) ) for top hit SNPs from GWAS results. These SNPs are also annotated in the manhattan plot in the HTML report.</p>
</li>
<li>
<p>Set <code>rarevar_min_log10p</code> to the min value ( -log10(pval) ) for top hit genes from rare variants analysis. These genes are also annotated in the general manhattan plot in the HTML report.</p>
</li>
<li>
<p>Set <code>clump_p1</code> to the maximum pvalue allowed for index SNPs during plink clumping to define top loci</p>
</li>
<li>
<p>If you are analyzing many phenotypes on a large dataset, we suggest to set <code>make_report</code> to false or at least disable the generation of the locus zoom plots by setting <code>n_top_loci_plot</code> to zero. Generating the HTML graphical report and especially the regional plots, can add a considerable amount of time when many phenotypes are tested together on a large dataset.</p>
</li>
<li>
<p>If you have categorical covariates, the maximum number of allowed categories is set to 10 bu default. You can adjsut this using the <code>maxCatLevels</code> parameter.</p>
</li>
</ul>
<h2 id="input-files">Input files</h2>
<p>Set the input files parameters as described in the input files sections.</p>
<ul>
<li>The <a href="../input-full-data/">full genetic dataset</a> to perform GWAS analysis</li>
<li>The <a href="../input-indep-snps/">reduced SNPs dataset</a> to perform REGENIE step1 regression</li>
<li>The <a href="input-rarevars-data.md">rare variants dataset and annotations</a> to perform rare variant tests</li>
<li>An optional set of files representing your <a href="../input-ld-panel/">LD panel</a> to perform loci clumping</li>
</ul>
<h2 id="multi-models-run">Multi-models run</h2>
<p>In case you want to configure a you also need to set the following parameters:</p>
<ul>
<li>
<p>Set <code>models_table</code> to a tab-separated file defining the models to test</p>
</li>
<li>
<p>Set <code>missing_tolerance</code> to the maximum allowed fraction of missing phenotype values when collecting uniform group of phenotypes for a run.</p>
</li>
</ul>
<h2 id="multi-projects-run">Multi-projects run</h2>
<p>In case you want to configure a multi-projects run you also need to set the following parameters:</p>
<ul>
<li>Set <code>projects_table</code> to a tab-separated file defining the projects to test</li>
</ul>
<h2 id="conditional-and-interaction-analysis">Conditional and interaction analysis</h2>
<p>In case you want to perform conditional or interaction analysis, you can configure the following parameters:</p>
<ul>
<li><code>interaction_cov</code>: run GxE test specifying the interacting covariate from covariate table</li>
<li><code>interaction_snp</code>: run GxG test specifying the interacting variant ID</li>
<li><code>condition_list</code>: run conditional analysis specifying a file with variant IDs to condition on</li>
</ul>
<p>Note that to perform conditional/interaction analysis, an additional genotype dataset must be provided containing the SNP(s) used for conditioning/interaction. This can be configured using the following parameters:</p>
<ul>
<li><code>additional_geno_file</code>: prefix of the genotype dataset containing vars in condition_list or interaction var. This is mandatory for conditional or interaction analysis</li>
<li><code>additional_geno_format</code>: can be bgen, pgen or bed.</li>
</ul>
<p>See the <a href="../input-conditional-analysis/">conditional analysis section</a> for more details.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      This pipeline is licensed under the MIT license
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.1e8ae164.min.js"></script>
      
    
  </body>
</html>