---
title: "Report: `r params$project`"
output:
  rmdformats::robobook:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: false
    highlight: tango
params:
  project: test-gwas
  date: 2021-08-09
  version: v0.2
  regenie_merged: Q1.regenie.gz
  regenie_filename: Q1.regenie.gz
  phenotype: Q1
  covariates: SEX
  phenotype_file: phenos_quant.pheno.validated.txt
  regenie_step1_log: test-ukbb-quant.step1.log
  regenie_step2_log: test-ukbb-quant.step2.log
  phenotype_log: phenos_quant.pheno.validated.log
  covariate_log: covars.cov.validated.log
  plot_ylimit: 0
  manhattan_annotation_enabled: TRUE
  annotation_min_log10p: 6
  annotated_tophits_filename: Q1.regenie.filtered.annotated.txt.gz
  annotated_toploci_filename: Q1.toploci.annot.tsv
---

```{r setup, include=FALSE}
library(DT)
library(data.table)
library(R.utils)
library(ggplot2)
library(skimr)
library(kableExtra)
library(dplyr)
library(ggfastman)
```

## Project Summary

| Parameter     | Value                                      |
| ------------- |--------------------------------------------|
| Project       | `r params$project`                         |
| Pipeline Version       | `r params$version`                |
| Date       | `r params$date`                         |
| Phenotype File      | `r params$phenotype_file`            |
| Phenotype       | `r params$phenotype`                     |
| Covariates       | `r params$covariates`                     |
| Regenie Output      | `r params$regenie_filename`             |

## Phenotype Statistics

### Overview

```{r, echo=FALSE}
phenotypeTable <- read.table(params$phenotype_file, header=TRUE, sep="\t", dec = ".")
kable(skim(phenotypeTable) %>%
  dplyr::filter(skim_variable == params$phenotype)) %>%
  kableExtra::scroll_box(width = "100%")
```

### Histogram
```{r, echo=FALSE}
ggplot(phenotypeTable, aes_string(x=params$phenotype)) +
  geom_histogram(color="black", fill="white",bins=30)
```

## Manhattan Plot

```{r, echo=FALSE, message=FALSE, warning=FALSE, time_it=TRUE, dpi=300}
#sources: https://www.r-graph-gallery.com/101_Manhattan_plot.html with modifications

regenieTable <- fread(params$regenie_merged, select = c("CHROM","GENPOS","LOG10P","ID"), header=TRUE, tmpdir="$PWD")
setnames(regenieTable, c("CHROM","GENPOS","LOG10P","ID"), c("chr","pos","pvalue","rsid")) 
regenieTable <- regenieTable[, pvalue := 10^(-pvalue)]
#source: https://www.r-graph-gallery.com/101_Manhattan_plot.html with modifications

annotatedTopHits <- fread(params$annotated_tophits_filename, select = c("ID", "CLOSEST_GENE_NAME"), header=TRUE, tmpdir="$PWD")
setnames(annotatedTopHits, "ID", "rsid")
#merge gene names of tophits to regenieTable

annotatedTopHits <- annotatedTopHits %>%  
    mutate(rsid = as.character(rsid))
regenieTable <- regenieTable %>%  
    mutate(rsid = as.character(rsid))

regenieTable_annotated <- left_join(regenieTable, annotatedTopHits, by = "rsid")

p <-  fast_manhattan(regenieTable_annotated, build='hg19', speed = "f",
                     color1 = "grey", color2 = "orange", pointsize = 2, 
                     highlight = regenieTable_annotated[regenieTable_annotated$pvalue < 5e-8,]$rsid) +
  geom_hline(yintercept = -log10(5e-08),
             linetype ="longdash",
             color ="firebrick") + # genomewide significance
  geom_hline(yintercept = -log10(1e-5),
             linetype ="longdash",
             color ="darkgrey")  + # suggestive significance

    # Custom the theme:
    theme_bw() +
    theme(
      legend.position="none",
      panel.border = element_blank(),
      axis.text = element_text(size = 12,
                               color = "black"),
      axis.title = element_text(size = 14),
      axis.ticks = element_line(color = "black")
      )

# custom y axis step 2 (depends if ymax is defined)

 if(params$plot_ylimit == 0) {
  p2 <- p + scale_y_continuous(expand = c(0, 0),
                       name=expression(-log[10](italic(P))),
                       limits = c(0, sum(max(-log10(regenieTable$pvalue)), 0.5)))
} else {
  p2 <- p + scale_y_continuous(expand = c(0, 0),
                       name=expression(-log[10](italic(P))),
                       limits = c(0, params$plot_ylimit))
}

rm(p)

 if(!params$manhattan_annotation_enabled) {
  p3 <- p2
} else {
  p3 <- p2 + ggrepel::geom_text_repel(data = . %>% filter(pvalue <= 5e-08) %>%
                                        group_by(CLOSEST_GENE_NAME) %>% 
                             top_n(1, -pvalue) %>% # extract highest y values
                             slice(1), # if there are ties, choose the first one
                             aes(label=CLOSEST_GENE_NAME), color =1)
}
 rm(regenieTable_annotated)
 rm(p2)
 
 p3

```

## QQ Plot

```{r, echo=FALSE, time_it = TRUE, results='asis'}
if(params$plot_ylimit == 0) {
  limit = NULL
} else {
  limit = c(0, params$plot_ylimit )
    cat(paste("Please note that the Y-axis limit has been set to '", params$plot_ylimit,"'. Please use the 'plot_ylimit' parameter to adapt this behaviour."))
}

fast_qq(regenieTable$pvalue, pointsize = 2)

rm(regenieTable)
```

## Top SNPs

```{r, echo=FALSE, message=FALSE, warning=FALSE}

annotatedTopHits <- fread(params$annotated_tophits_filename, select = c("CHROM", "GENPOS", "LOG10P", "ID", "CLOSEST_GENE_NAME", "CLOSEST_GENE_CHROMOSOME", "CLOSEST_GENE_START", "CLOSEST_GENE_END", "CLOSEST_GENE_DISTANCE"), header=TRUE, tmpdir="$PWD")

topHitsGroupedByGene <-
  annotatedTopHits %>%
  group_by(CLOSEST_GENE_NAME, CLOSEST_GENE_CHROMOSOME, CLOSEST_GENE_START, CLOSEST_GENE_END) %>%
  summarise(GENE_LOG10P = max(LOG10P), VARIANTS = n()) %>%
  arrange(desc(GENE_LOG10P))

datatable(topHitsGroupedByGene)

```

## Top Loci

```{r, echo=FALSE, message=FALSE, warning=FALSE}
if (file.exists(params$annotated_toploci_filename)) {
  annotatedTopLoci <- fread(params$annotated_toploci_filename, header=TRUE, tmpdir="$PWD")
  annotatedTopLoci <- annotatedTopLoci[order(annotatedTopLoci$P),]
  
  datatable(annotatedTopLoci)
} else {
  cat("No clumping loci available")
}
```

## Validation and Logs

### Phenotype File Validation

```{r, echo=FALSE}
phenotype_log <- read.table(params$phenotype_log,
                      sep ="\t", header = TRUE, dec =".")
kable(phenotype_log)
```

### Covariate File Validation

```{r, echo=FALSE, results='asis'}
if(file.exists(params$covariate_log)) {
step1_log <- read.table(params$covariate_log,sep ="\t", header = TRUE, dec =".")
kable(step1_log)
} else {
    cat("*No covariates available.*")
}
```

### Regenie Step 1 Log

```{r, echo=FALSE, results='asis'}
if(file.exists(params$regenie_step1_log)) {
step1_log <- read.table(params$regenie_step1_log,sep ="\t", header = TRUE, dec =".")
kable(step1_log)
} else {
    cat("*Regenie step 1 skipped.*")
}
```

### Regenie Step 2 Log

```{r, echo=FALSE}
step2_log <- read.table(params$regenie_step2_log,
                      sep ="\t", header = TRUE, dec =".")
kable(step2_log)
```


---

<small>
This report has been created with **[nf-fast-regenie `r params$version`](https://github.com/genepi/nf-gwas)**, a nextflow pipeline developed by [Edoardo Giacopuzzi](mailto:edoardo.giacopuzzi@fht.org) at the Human Technopole Foundation, Milan, Italy. 
</small>
