
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Edoardi Giacopuzzi">
      
      
      
        <link rel="prev" href="../top-hits-annotation/">
      
      
        <link rel="next" href="../outputs/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.17">
    
    
      
        <title>full list of params - nf-regenie-gwas</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.bcfcd587.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue" data-md-color-accent="blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#full-list-of-parameters" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="nf-regenie-gwas" class="md-header__button md-logo" aria-label="nf-regenie-gwas" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            nf-regenie-gwas
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              full list of params
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/HTGenomeAnalysisUnit/nf-pipeline-regenie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../quick-start/" class="md-tabs__link">
        
  
    
  
  Quick start

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../running-modes/" class="md-tabs__link">
          
  
  Running mode

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../input-phenotype-file/" class="md-tabs__link">
          
  
  Input files

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../main-parameters/" class="md-tabs__link">
          
  
  Configuration

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../outputs/" class="md-tabs__link">
        
  
    
  
  Output files

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../additional-notes/" class="md-tabs__link">
        
  
    
  
  Additional notes

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="nf-regenie-gwas" class="md-nav__button md-logo" aria-label="nf-regenie-gwas" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    nf-regenie-gwas
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/HTGenomeAnalysisUnit/nf-pipeline-regenie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../quick-start/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick start
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Running mode
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Running mode
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../running-modes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    single model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../running-modes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    multi projects
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../running-modes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    multi models
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Input files
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Input files
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-phenotype-file/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Phenotype / Covariate file
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-full-data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Full genetic dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-indep-snps/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Independent QCed SNPs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-ld-panel/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD panel (optional)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-rare-variants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inputs for rare variants tests
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-models-table/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Models table (multi models run)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-projects-table/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Project table (multi projects run)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-conditional-analysis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Conditional / interaction analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-reuse-step1/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Re-use step1 predictions
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Configuration
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../main-parameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    main parameters
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tower-monitoring/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    tower monitoring
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../hpc-profile/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    define HPC profile
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../parallelization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    optimize parallelization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../top-hits-annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    top hits annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    full list of params
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    full list of params
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#global-settings-of-resources" class="md-nav__link">
    <span class="md-ellipsis">
      Global settings of resources
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#general-settings-per-run" class="md-nav__link">
    <span class="md-ellipsis">
      General settings per run
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multi-model-specific-settings" class="md-nav__link">
    <span class="md-ellipsis">
      Multi-model specific settings
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Input files
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Input files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genetic-data" class="md-nav__link">
    <span class="md-ellipsis">
      Genetic data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phenotypes" class="md-nav__link">
    <span class="md-ellipsis">
      Phenotypes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#covariates" class="md-nav__link">
    <span class="md-ellipsis">
      Covariates
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#rare-variants-accessory-files" class="md-nav__link">
    <span class="md-ellipsis">
      Rare variants accessory files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-optional-inputs" class="md-nav__link">
    <span class="md-ellipsis">
      Other optional inputs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step1-pre-processing-settings" class="md-nav__link">
    <span class="md-ellipsis">
      STEP1 PRE-PROCESSING SETTINGS
    </span>
  </a>
  
    <nav class="md-nav" aria-label="STEP1 PRE-PROCESSING SETTINGS">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#snp-pruning" class="md-nav__link">
    <span class="md-ellipsis">
      SNP pruning
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#regenie-step1-settings" class="md-nav__link">
    <span class="md-ellipsis">
      REGENIE STEP1 SETTINGS
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#regenie-step2-settings" class="md-nav__link">
    <span class="md-ellipsis">
      REGENIE STEP2 SETTINGS
    </span>
  </a>
  
    <nav class="md-nav" aria-label="REGENIE STEP2 SETTINGS">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#general-settings" class="md-nav__link">
    <span class="md-ellipsis">
      General settings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vcf-to-pgen-conversion-settings" class="md-nav__link">
    <span class="md-ellipsis">
      VCF to PGEN conversion settings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gwas-analysis-settings" class="md-nav__link">
    <span class="md-ellipsis">
      GWAS analysis settings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#rare-variant-analysis-settings" class="md-nav__link">
    <span class="md-ellipsis">
      Rare variant analysis settings
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gwas-results-annotation-and-clumping" class="md-nav__link">
    <span class="md-ellipsis">
      GWAS RESULTS ANNOTATION AND CLUMPING
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rare-variants-results-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      RARE VARIANTS RESULTS ANNOTATION
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#report-settings" class="md-nav__link">
    <span class="md-ellipsis">
      REPORT SETTINGS
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../outputs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Output files
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../additional-notes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Additional notes
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#global-settings-of-resources" class="md-nav__link">
    <span class="md-ellipsis">
      Global settings of resources
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#general-settings-per-run" class="md-nav__link">
    <span class="md-ellipsis">
      General settings per run
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multi-model-specific-settings" class="md-nav__link">
    <span class="md-ellipsis">
      Multi-model specific settings
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Input files
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Input files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genetic-data" class="md-nav__link">
    <span class="md-ellipsis">
      Genetic data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phenotypes" class="md-nav__link">
    <span class="md-ellipsis">
      Phenotypes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#covariates" class="md-nav__link">
    <span class="md-ellipsis">
      Covariates
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#rare-variants-accessory-files" class="md-nav__link">
    <span class="md-ellipsis">
      Rare variants accessory files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-optional-inputs" class="md-nav__link">
    <span class="md-ellipsis">
      Other optional inputs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step1-pre-processing-settings" class="md-nav__link">
    <span class="md-ellipsis">
      STEP1 PRE-PROCESSING SETTINGS
    </span>
  </a>
  
    <nav class="md-nav" aria-label="STEP1 PRE-PROCESSING SETTINGS">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#snp-pruning" class="md-nav__link">
    <span class="md-ellipsis">
      SNP pruning
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#regenie-step1-settings" class="md-nav__link">
    <span class="md-ellipsis">
      REGENIE STEP1 SETTINGS
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#regenie-step2-settings" class="md-nav__link">
    <span class="md-ellipsis">
      REGENIE STEP2 SETTINGS
    </span>
  </a>
  
    <nav class="md-nav" aria-label="REGENIE STEP2 SETTINGS">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#general-settings" class="md-nav__link">
    <span class="md-ellipsis">
      General settings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vcf-to-pgen-conversion-settings" class="md-nav__link">
    <span class="md-ellipsis">
      VCF to PGEN conversion settings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gwas-analysis-settings" class="md-nav__link">
    <span class="md-ellipsis">
      GWAS analysis settings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#rare-variant-analysis-settings" class="md-nav__link">
    <span class="md-ellipsis">
      Rare variant analysis settings
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gwas-results-annotation-and-clumping" class="md-nav__link">
    <span class="md-ellipsis">
      GWAS RESULTS ANNOTATION AND CLUMPING
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rare-variants-results-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      RARE VARIANTS RESULTS ANNOTATION
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#report-settings" class="md-nav__link">
    <span class="md-ellipsis">
      REPORT SETTINGS
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="full-list-of-parameters">Full list of parameters</h1>
<h2 id="global-settings-of-resources">Global settings of resources</h2>
<p>Global settings for max resources available per process in your system</p>
<ul>
<li>max_memory </li>
<li>max_cpus   </li>
<li>max_time   </li>
</ul>
<h2 id="general-settings-per-run">General settings per run</h2>
<ul>
<li>project: project name. <strong>Max 50 chars. No spaces allowed. Use only alphanumeric, underscores, or dashes</strong></li>
<li>project_date: usually configured automatically to execution date, but you can override the date manually</li>
<li>outdir: a specific outdir for the project, if null the project id is used as output dir, otherwise <code>outdir/project_id/</code></li>
<li>chromosomes: chromosomes to include in the analysis. Accepts comma separated lists and intervals like <code>1-5,8-12,22</code></li>
<li>master_log_dir: set this to redirect all logs to a specific folder</li>
</ul>
<h2 id="multi-model-specific-settings">Multi-model specific settings</h2>
<pre><code>models_table                          = null //when a table is provided, this activates master table mode
pheno_chunk_size                      = 50 //maximum number of phenotypes for a single GWAS run
missing_tolerance                     = 0.1 //max fraction of missing values per phenotype when assembling execution groups
</code></pre>
<h2 id="input-files">Input files</h2>
<h3 id="genetic-data">Genetic data</h3>
<pre><code>genotypes_array                       = null //genotypes data for step 1. Can be bgen, pgen, bed or vcf.gz
genotypes_imputed                     = null //genotype data for gwas analysis. Can be bgen, pgen, bed or vcf.gz
genotypes_rarevar                     = null //genotype data for rare variants analysis. Can be bgen, pgen, plink bed or vcf.gz
genotypes_build                       = null //genome build
genotypes_imputed_format              = null //input data format for gwas. Can be bgen, pgen, bed or vcf
genotypes_rarevar_format              = null //input data format for rare variants. Can be bgen, pgen, bed or vcf
imputed_sample_file                   = 'NO_SAMPLE_FILE' //Provide a specific sample file to use with gwas bgen input
rarevar_sample_file                   = 'NO_SAMPLE_FILE' //Provide a specific sample file to use with rare variants bgen input
ld_panel                              = 'NO_LD_FILE' //A pattern pointing to a subset of genotypes to be used for LD computation. Optional, but highly reccomended for large datasets
</code></pre>
<h3 id="phenotypes">Phenotypes</h3>
<pre><code>phenotypes_filename                   = null //table of phenos - required
phenotypes_columns                    = null //comma separated list of col names in pheno file to be analyzed
phenotypes_binary_trait               = null //true for binary
</code></pre>
<h3 id="covariates">Covariates</h3>
<pre><code>covariates_filename                   = 'NO_COV_FILE' //files containing covariates, use NO_COV_FILE when absent
covariates_columns                    = '' //comma-separated list of covariates column names 
covariates_cat_columns                = '' //comma-separated list of categorical covariate column names
maxCatLevels                          = 10 //maximum number of allowed levels for categorical covars
</code></pre>
<h3 id="rare-variants-accessory-files">Rare variants accessory files</h3>
<p>These are mandatory when running rare var analysis</p>
<pre><code>rarevar_set_list_file                 = null //set list file as defined in regenie docs
rarevar_anno_file                     = null //variant annotation file as defined in regenie docs
rarevar_mask_file                     = null //mask definition file as defined in regenie docs
</code></pre>
<h3 id="other-optional-inputs">Other optional inputs</h3>
<pre><code>genes_bed                             = false //an optional .bed file specifying genes intervals used for results annotation
genes_ranges                          = false //an optional .interval file specifying genes intervals used for loci annotation
</code></pre>
<h2 id="step1-pre-processing-settings">STEP1 PRE-PROCESSING SETTINGS</h2>
<h3 id="snp-pruning">SNP pruning</h3>
<pre><code>prune_enabled                         = false
prune_maf                             = 0.01
prune_window_kbsize                   = 1000
prune_step_size                       = 100
prune_r2_threshold                    = 0.9
</code></pre>
<h3 id="filtering">Filtering</h3>
<pre><code>qc_maf                                = '0.01'
qc_mac                                = '100'
qc_geno                               = '0.05'
qc_hwe                                = '1e-15'
qc_mind                               = '0.1'
</code></pre>
<h2 id="regenie-step1-settings">REGENIE STEP1 SETTINGS</h2>
<pre><code>regenie_bsize_step1                   = 1000
regenie_premade_predictions           = false //or pattern to regenie step1 files
save_step1_predictions                = true
regenie_force_step1                   = false
regenie_ref_first_step1               = false
step1_use_loocv                       = false
step1_niter                           = 30
step1_n_chunks                        = 100 // N chunks when performing step1 L0 regression
</code></pre>
<h2 id="regenie-step2-settings">REGENIE STEP2 SETTINGS</h2>
<h3 id="general-settings">General settings</h3>
<pre><code>phenotypes_delete_missings            = false //remove samples with missing data at any of the phenotypes
regenie_bsize_step2                   = 400
regenie_ref_first_step2               = true
regenie_skip_predictions              = false //skip reading the step1 predictions (corresponds to simple linear/logistic regression)
regenie_range                         = '' // when splitting is not active you can use this to specify a genomic range for step2 analysis
regenie_extract_snps                  = '' // when splitting is not active you can specify a file containing a list of variant IDs to restrict step2 analysis
regenie_extract_genes                 = '' // when splitting is not active you can use this to specify a file containing a list of genes to restrict step2 analysis
interaction_cov                       = null // run GxE test in GWAS specifying the interacting covariate from covariate table
interaction_snp                       = null // run GxG test in GWAS specifying the interacting variant ID
condition_list                        = null // run conditional analysis in GWAS specifying a files with variant IDs to condition on
additional_geno_file                  = null // prefix of the genotype dataset containing vars in condition_list or interaction var. This is mandatory for conditional or interaction analysis
additional_geno_format                = null // can be bgen, pgen or bed.
save_chunks_file                      = true
save_step2_logs                       = true
save_bgen_index                       = true
save_bgen_sample                      = true    
save_converted_bgen                   = true
</code></pre>
<h3 id="vcf-to-pgen-conversion-settings">VCF to PGEN conversion settings</h3>
<pre><code>gwas_read_dosage_from                 = 'DS' //DS (usually for VCF from imputateion) or GP (usually VCF from sequencing)
rarevar_read_dosage_from              = 'GP' //DS (usually for VCF from imputateion) or GP (usually VCF from sequencing)
import_dosage_certainty               = 0.7 //when using GP, the certainty threshold to import dosages. If none of the probabilities is above this, the genotype is set to missing
vcf_fixed_fid                         = null //when null vcf is converted to pgen using --double-id, otherwise fid is fixed to this value
</code></pre>
<h3 id="gwas-analysis-settings">GWAS analysis settings</h3>
<pre><code>step2_gwas_split                      = true //when true activate split of step2 by variant chunks
step2_gwas_chunk_size                 = 100000 //n variants per chunk when running gwas in split mode
regenie_gwas_test                     = 'additive' //additive, dominant, recessive
regenie_min_imputation_score          = '0.00'
regenie_gwas_min_mac                  = '50' // min MAC for variants to be included in step2 for gwas 
regenie_firth                         = true
regenie_firth_approx                  = true
</code></pre>
<h3 id="rare-variant-analysis-settings">Rare variant analysis settings</h3>
<pre><code>step2_rarevar_split                   = true //when true activate split of step2 by gene chunks
step2_rarevar_chunk_size              = 200 //n genes per chunk when running rare variant test in split mode
regenie_rarevar_min_mac               = '1' // min MAC for variants to be included in step2 for rare vars
rarevars_aaf_bins                     = '0.01,0.05' //comma-separated list of AAF upper bounds to use when building masks for burden test
rarevars_vc_test                      = 'skat,skato,acatv,acato' //comma-separated list of SKAT/ACAT-type tests to run
rarevars_joint_test                   = 'minp,acat' //comma-separated list of joint tests to run. Choices: minp,acat,sbat. Note that sbat can't run on binary phenotypes
rarevars_vc_maxAAF                    = '0.05' //AAF upper bound to use for SKAT/ACAT-type tests
regenie_build_mask                    = 'max' //build mask for rare variant test. Can be max, sum, or a comphet
rarevars_write_mask_snplist           = false //when true write list of variants that went into each mask to file
</code></pre>
<h2 id="gwas-results-annotation-and-clumping">GWAS RESULTS ANNOTATION AND CLUMPING</h2>
<pre><code>genes_group                           = 'protein_coding' //genes group to use for annotation. Can be all or protein_coding
annotation_min_log10p                 = 7.3 //results with -log10(p) above this will be reported as top hits with annotated genes
annotation_interval_kb                = 25
clumping                              = true
clump_p1                              = 5e-8
clump_p2                              = 1e-4
clump_kb                              = 250
</code></pre>
<h2 id="rare-variants-results-annotation">RARE VARIANTS RESULTS ANNOTATION</h2>
<pre><code>rarevar_min_log10p                    = 5 //results with -log10(p) above this will be reported as top hits
rarevar_stat_test                     = "BONF_bygroup" //Stat value to filter on. Possible values: "FDR_bygroup", "FDR_alltests", "BONF_bygroup", "BONF_alltests"
rarevar_stat_test_threshold           = 1.3 //results with -log10(stat_value) above this will be annotated in a dedicated manhattan plot
</code></pre>
<h2 id="report-settings">REPORT SETTINGS</h2>
<pre><code>make_report                           = true //it can be useful to disable when analyzing many phenotypes
manhattan_annotations                 = 'genes' //how to annotate peaks in manhattan plot. Either 'genes' or 'snpid'
regional_plot_window_kb               = 300 //window size in kb for regional plots. This value is added on each side of the locus when plotting
n_top_loci_plot                       = 5 //number of top loci to plot in regional plots
</code></pre>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      This pipeline is licensed under the MIT license
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.1e8ae164.min.js"></script>
      
    
  </body>
</html>